### PR TITLE
cherry-pick of #63495: kubeadm - fix upgrades with external etcd

### PR DESCRIPTION
**Pull request by detiber**
**What this PR does / why we need it**:

Backports fixes for kubeadm upgrades with external etcd

**Release note**:
```release-note
kubeadm - fix upgrades with external etcd
```

### CODE REVIEW

**Comment by MaciekPytel** [comment]
@detiber This needs to merge by tomorrow EOD to make it into 1.10.3.

File: cmd/kubeadm/app/cmd/upgrade/plan_test.go [hunk 1]
```diff
@@ -62,16 +62,27 @@ func TestSortedSliceFromStringIntMap(t *testing.T) {
...
+		name          string
 		upgrades      []upgrade.Upgrade
 		buf           *bytes.Buffer
 		expectedBytes []byte
+		externalEtcd  bool
 	}{
 		{
+			name:     "Up to date",
 			upgrades: []upgrade.Upgrade{},
 			expectedBytes: []byte(`Awesome, you're up-to-date! Enjoy!
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/upgrade/plan_test.go]: huh, i wonder how the old code passed `go fmt`. it complains for me when not indenting the members property.

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 2]
```diff
@@ -83,40 +86,127 @@ FlagLoop:
...
+	}
+	for _, v := range versions {
+		if clusterVersion == "" {
+			// This is the first version we've seen
+			clusterVersion = v
+		} else if v != clusterVersion {
+			return "", fmt.Errorf("etcd cluster contains endpoints with mismatched versions: %v", versions)
+		} else {
+			clusterVersion = v
+		}
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go]: change only if see fit :+1: 
```
for _, v := range versions {
	if clusterVersion != "" && clusterVersion != v {
		return "", fmt.Errorf("etcd cluster contains endpoints with mismatched versions: %v", versions)
	}
	clusterVersion = v
}
```

File: cmd/kubeadm/app/cmd/upgrade/plan_test.go [hunk 3]
```diff
@@ -330,19 +345,71 @@ Note: Before you can perform this upgrade, you have to update kubeadm to v1.9.0-
...
+					After: upgrade.ClusterState{
+						KubeVersion:    "v1.9.3",
+						KubeadmVersion: "v1.9.3",
+						DNSVersion:     "1.14.8",
+						EtcdVersion:    "3.1.12",
+					},
+				},
+			},
+			externalEtcd: true,
+			expectedBytes: []byte(`External components that should be upgraded manually before you upgrade the control plane with 'kubeadm upgrade apply':
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/upgrade/plan_test.go]: these byte comparisons without helper functions are scary, but we have them in the old test code, so :+1:

**Comment by stealthybox** [comment]
Ran into this when upgrading staticpod etcd:
```
root@vagrant:~# /vagrant/bin/cherry-63925_kubeadm upgrade apply v1.10.2
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade/version] You have chosen to change the cluster version to "v1.10.2"
[upgrade/versions] Cluster version: v1.9.7
[upgrade/versions] kubeadm version: v1.10.3-beta.0.50+29974a37bba8ad
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler]
[upgrade/apply] Upgrading your Static Pod-hosted control plane to version "v1.10.2"...
Static pod: kube-apiserver-vagrant hash: 8995194791ca1350e37e54cdb9db6c44
Static pod: kube-controller-manager-vagrant hash: 305b23ead04fc3114ba6755a75d9d17d
Static pod: kube-scheduler-vagrant hash: 96944ce896a1ba4844bab386f40c0acc
[upgrade/etcd] Upgrading to TLS for etcd
[upgrade/apply] FATAL: etcd cluster is not healthy: etcdclient: no available endpoints
```
Are you able to reproduce?

I was able to upgrade an insecure external etcd successfully.

**Review by stealthybox — CHANGES_REQUESTED** [review]
@detiber -- I'm still very confused, but I was able to fix this problem with a small patch:

```diff
diff --git a/cmd/kubeadm/app/cmd/upgrade/apply.go b/cmd/kubeadm/app/cmd/upgrade/apply.go
index b7fc2c0443..ad6b420a73 100644
--- a/cmd/kubeadm/app/cmd/upgrade/apply.go
+++ b/cmd/kubeadm/app/cmd/upgrade/apply.go
@@ -35,7 +35,7 @@ import (
        "k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
        configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
        dryrunutil "k8s.io/kubernetes/cmd/kubeadm/app/util/dryrun"
-       etcdutil "k8s.io/kubernetes/cmd/kubeadm/app/util/etcd"
+       // etcdutil "k8s.io/kubernetes/cmd/kubeadm/app/util/etcd"
        "k8s.io/kubernetes/pkg/api/legacyscheme"
        "k8s.io/kubernetes/pkg/util/version"
 )
@@ -269,9 +269,8 @@ func PerformStaticPodUpgrade(client clientset.Interface, waiter apiclient.Waiter
                return err
        }

-       // These are uninitialized because passing in the clients allow for mocking the client during testing
-       var oldEtcdClient, newEtdClient etcdutil.Client
-       return upgrade.StaticPodControlPlane(waiter, pathManager, internalcfg, etcdUpgrade, oldEtcdClient, newEtdClient)
+       return upgrade.StaticPodControlPlane(waiter, pathManager, internalcfg, etcdUpgrade, nil, nil)
+       // The arguments, oldEtcdClient and newEtdClient, are uninitialized because passing in the clients allow for mocking the client during testing
 }

 // DryRunStaticPodUpgrade fakes an upgrade of the control plane
```

[original source:  app/cmd/upgrade/apply.go#L273-L274](https://github.com/detiber/kubernetes/blob/29974a37bba8ad7e247d0a57386ce1eb65ce54a1/cmd/kubeadm/app/cmd/upgrade/apply.go#L273-L274)

The client vars were previously being initialized with a zero-value (not nil).
This actually made the very important comparisons for those clients against `nil` unreachable from the cmd entrypoint.

I'm not sure how initialization and method behavior for these clients was ever working now that I've found this bug, but I'm going to take a break.

We definitely need to validate the behavior in all of the common cases again.

**Comment by MaciekPytel** [comment]
Added the milestone, so this moves faster through submit queue and make it before branch freeze.

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 2 above)
Comment (detiber) [reply cmd/kubeadm/app/util/etcd/etcd.go]: I added this to my PR for fixing the static pod upgrade regression in master.

**Review by stealthybox — APPROVED** [review]
ty @detiber 🌮 
I've verified locally that this now works in both cases.
/lgtm

**Review by timothysc — COMMENTED** [review]
/approve

**Pull request merged**